fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java

```diff
@@ -139,8 +167,114 @@ public CollectionData getOverdueCollectionData(final Loan loan) {
         if (delinquentDays > 0) {
             collectionData.setDelinquentDays(delinquentDays);
         }
+        return new LoanDelinquencyData(collectionData, loanInstallmentsCollectionData);
+    }
 
-        log.debug("Result: {}", collectionData.toString());
+    private CollectionData getInstallmentOverdueCollectionData(final Loan loan, final LoanRepaymentScheduleInstallment installment) {
```

Comment:
There are a lot of new code, but i am not really see why?

Reply:
Merged logic of calculation loan and installments delinquency into one method -> getLoanDelinquencyData,

In this method iterating to all loan installments and calculating delinquency for each installment in ->getInstallmentOverdueCollectionData 

If installment is not paid 

checking that
[1] If installment is overdue calculating delinquent data with considering any chargeback transactions  in -> calculateDelinquencyDataForOverdueInstallment method
[2[ If installment is not overdue calculating delinquent data if there is chargeback transaction before installment due date and current business date in ->calculateDelinquencyDataForNonOverdueInstallment method

If installment is paid setting delinquent days as 0. 

Adding installments delinquency calculation to final calculation data if configured.

For calculating loan delinquency taking overdueSince date for oldest overdue installment or first non due installment with chargeback transaction which is before installment due date and current business date.


getLoanDelinquencyData returns LoanDelinquencyData which has loanCollectionData and map of installment Collection Data.